### PR TITLE
MODFEE-29 Double to MonetaryValue

### PR DESCRIPTION
**Pull request by nikolai991**
Resolves https://issues.folio.org/browse/MODFEE-29.
### Description
All of the money fields' types should be changed to BigDecimal because Double can't accurately represent money values.
### Approach 
During the task resolution, the types of the objects' properties representing money values have been changed to  org.folio.rest.domain.MonetaryValue in json schemas. The MonetaryValue class is serialized to one field in a json object using MonetaryValueSerializer (JsonSerializer implementation), which is located in JsonHelper.java.

### CODE REVIEW

File: src/test/java/org/folio/rest/impl/AccountsRefundAPITests.java [hunk 1]
```diff
@@ -460,23 +513,37 @@ public void bulkRefundAmountIsDistributedBetweenAccountsEvenlyRecursively() {
...
+   MonetaryValue amount = new MonetaryValue(5.0);
+    DefaultBulkActionRequest request = createBulkRequest(amount, TWO_ACCOUNT_IDS);
+    testBulkRefundFailure(initialAmount, payAmount, transferAmount, waiveAmount, request, SC_UNPROCESSABLE_ENTITY, ERROR_MESSAGE,
       TWO_ACCOUNT_IDS);
   }
 
   @Test
   public void bulkReturn404WhenAccountDoesNotExist() {
     String errorMessageTemplate = "Fee/fine ID %s not found";
+    MonetaryValue amount = new MonetaryValue( 10.0);
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/AccountsRefundAPITests.java:544]: Remove

File: src/test/java/org/folio/rest/impl/AccountsRefundAPITests.java [hunk 2]
```diff
@@ -599,19 +684,19 @@ public void bulkRefundForTransfersToMultipleTransferAccounts() {
...
 
     verifyThatFeeFineBalanceChangedEventsWereSent(firstAccount, secondAccount);
   }
 
-  private void testSingleAccountRefundSuccess(double initialAmount, double payAmount, double transferAmount,
-    double waiveAmount, double requestedAmount, FeeFineStatus expectedStatus,
-    String expectedPaymentStatus, List<Matcher<JsonObject>> expectedFeeFineActions) {
+  private void testSingleAccountRefundSuccess(MonetaryValue initialAmount, MonetaryValue payAmount, MonetaryValue transferAmount,
+                                              MonetaryValue waiveAmount, MonetaryValue requestedAmount, FeeFineStatus expectedStatus,
+                                              String expectedPaymentStatus, List<Matcher<JsonObject>> expectedFeeFineActions) {
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/AccountsRefundAPITests.java:697]: Indentation

File: src/test/java/org/folio/rest/impl/CashDrawerReconciliationReportTest.java [hunk 3]
```diff
@@ -62,6 +64,34 @@
...
+        null,
+        null,
+        null,
+        null,
+        null,
+        null,
+        new MonetaryValue(7.0),
+        new MonetaryValue(8.0),
+        new MonetaryValue(9.0)
+      };
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/CashDrawerReconciliationReportTest.java:92]: Refactor

File: src/test/java/org/folio/rest/impl/FeeFineActionsAPITest.java [hunk 4]
```diff
@@ -464,8 +467,8 @@ private static Feefineaction createCharge(User user, Account account, boolean no
       .withNotify(notify)
       .withTypeAction("Overdue fine")
       .withDateAction(new Date())
-      .withAmountAction(Double.valueOf(ACTION_AMOUNT))
-      .withBalance(Double.valueOf(ACCOUNT_REMAINING))
+      .withAmountAction(new MonetaryValue(new BigDecimal(ACTION_AMOUNT)))
+      .withBalance(new MonetaryValue(new BigDecimal(ACCOUNT_REMAINING)))
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/FeeFineActionsAPITest.java:471]: use new MonetaryValue(string)

File: src/test/java/org/folio/rest/impl/RefundReportTest.java [hunk 5]
```diff
@@ -101,6 +102,42 @@
...
+              null,
+              null,
+              new MonetaryValue(17.0),
+              null,
+              null,
+              new MonetaryValue(20.0)
+         };
+
+  private  static  final  MonetaryValue[] BALANCE = {
+              new MonetaryValue(0.0),
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/RefundReportTest.java:130]: Refactor

File: src/test/java/org/folio/rest/impl/FeeFineReportsAPITestBase.java [hunk 6]
```diff
@@ -54,13 +54,13 @@ void removeLocaleSettingsStub() {
     removeStub(localeSettingsStubMapping);
   }
 
-  Account charge(String userID, Double amount, String feeFineType,
+  Account charge(String userID, MonetaryValue amount, String feeFineType,
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/FeeFineReportsAPITestBase.java:57]: Keep double in these methods, convert to MonetaryValue here

File: src/test/java/org/folio/rest/impl/RefundReportTest.java [hunk 7]
```diff
@@ -317,16 +354,25 @@ public void fullyRefundedTimeZoneTest() {
 
   @Test
   public void multiplePaymentsSameMethodFullyRefunded() {
-    Account account = charge(10.0, "ff-type", item1.getId());
+    Account account = charge(AMOUNT[10], "ff-type", item1.getId());
+
+    MonetaryValue amount1 = new MonetaryValue(3.1);
+    MonetaryValue balance1 = new MonetaryValue(6.9);
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/RefundReportTest.java:360]: Remove

File: src/test/java/org/folio/rest/impl/RefundReportTest.java [hunk 8]
```diff
@@ -509,35 +571,49 @@ public void multipleAccountMultipleRefunds() {
...
+
+    MonetaryValue amount2 = new MonetaryValue(3.2);
+    MonetaryValue balance2 = new MonetaryValue(3.7);
+
+    MonetaryValue balance3 = new MonetaryValue(5.7);
+
+    MonetaryValue balance4 = new MonetaryValue(20.0);
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/RefundReportTest.java:587]: Remove

File: src/test/java/org/folio/rest/impl/RefundReportTest.java [hunk 9]
```diff
@@ -658,15 +741,15 @@ private RefundReportEntry createResponseForMinimumViableData(ReportSourceObjects
...
+    String type, String method, MonetaryValue amount, MonetaryValue balance, String staffInfo,
     String patronInfo, String txInfo) {
 
     return createAction(USER_ID_1, actionCounter, account, dateTime, type, method, amount, balance,
       staffInfo, patronInfo, txInfo);
   }
 
   private Feefineaction createActionWithNullComments(Account account, String dateTime,
-    String type, String method, Double amount, Double balance, String txInfo) {
+                String type, String method, MonetaryValue amount, MonetaryValue balance, String txInfo) {
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/RefundReportTest.java:752]: Indentation

File: src/test/java/org/folio/rest/impl/RefundReportTest.java [hunk 10]
```diff
@@ -658,15 +741,15 @@ private RefundReportEntry createResponseForMinimumViableData(ReportSourceObjects
   }
 
   private Feefineaction createAction(int actionCounter, Account account, String dateTime,
-    String type, String method, Double amount, Double balance, String staffInfo,
+    String type, String method, MonetaryValue amount, MonetaryValue balance, String staffInfo,
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/RefundReportTest.java:744]: Keep double

File: src/test/java/org/folio/rest/impl/accountactionchecks/AccountsActionChecksAPITestsBase.java [hunk 11]
```diff
@@ -25,9 +27,9 @@
 
   protected static final String ACCOUNTS_TABLE = "accounts";
   protected static final String ERROR_MESSAGE_ALREADY_CLOSED = "Fee/fine is already closed";
-  protected static final double ACCOUNT_INITIAL_AMOUNT = 9.00;
-  protected static final double ACCOUNT_REMAINING_AMOUNT = 4.55;
-  protected static final double REQUESTED_AMOUNT = 1.23;
+  protected static final MonetaryValue ACCOUNT_INITIAL_AMOUNT = new MonetaryValue(9.00);
+  protected static final MonetaryValue ACCOUNT_REMAINING_AMOUNT = new MonetaryValue(4.55);
+  protected static final MonetaryValue REQUESTED_AMOUNT =new MonetaryValue( 1.23);
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/accountactionchecks/AccountsActionChecksAPITestsBase.java:32]: space

File: src/test/java/org/folio/rest/impl/accountactionchecks/AccountsActionChecksAPITestsBase.java [hunk 12]
```diff
@@ -67,20 +67,18 @@ protected void actionCheckAmountShouldNotBeAllowedWithExceededAmount(boolean bul
     ResourceClient actionCheckClient) {
 
     String expectedErrorMessage = "Requested amount exceeds remaining amount";
-    String amount = String.valueOf(bulk ? ACCOUNT_REMAINING_AMOUNT * 2 + 1 : REQUESTED_AMOUNT + 10);
+    String amount = String.valueOf(bulk ? ACCOUNT_REMAINING_AMOUNT.toDouble() * 2 + 1 : REQUESTED_AMOUNT.toDouble() + 10);
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/accountactionchecks/AccountsActionChecksAPITestsBase.java:70]: Calculate as MonetaryValue

File: src/test/java/org/folio/rest/impl/accountactionchecks/AccountsActionChecksAPITestsBase.java [hunk 13]
```diff
@@ -67,20 +67,18 @@ protected void actionCheckAmountShouldNotBeAllowedWithExceededAmount(boolean bul
...
 
-  protected void actionCheckRefundAmountShouldNotBeAllowedWithExceededAmount(boolean bulk,
-    ResourceClient actionCheckClient) {
-
+  protected void actionCheckRefundAmountShouldNotBeAllowedWithExceededAmount(boolean bulk, ResourceClient actionCheckClient) {
     String expectedErrorMessage =
       "Refund amount must be greater than zero and less than or equal to Selected amount";
 
-    String amount = String.valueOf(REQUESTED_AMOUNT + 10);
+    String amount = String.valueOf(REQUESTED_AMOUNT.toDouble() + 10);
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/accountactionchecks/AccountsActionChecksAPITestsBase.java:81]: Same here

File: src/test/java/org/folio/rest/impl/accountactionchecks/AccountsActionChecksAPITestsBase.java [hunk 14]
```diff
@@ -135,15 +133,15 @@ protected void actionCheckAmountShouldNotBeAllowedForClosedAccount(boolean bulk,
...
+        account.setRemaining(new MonetaryValue(new BigDecimal("0.00")));
         account.getStatus().setName(FeeFineStatus.CLOSED.getValue());
         accountsClient.update(account.getId(), account);
       });
   }
 
 
   protected void successfulActionCheckHandlesLongDecimalsCorrectly(ResourceClient client) {
-    firstAccount.setRemaining(1.235987654321); // will be rounded to 1.24
+    firstAccount.setRemaining(new MonetaryValue(new BigDecimal("1.235987654321"))); // will be rounded to 1.24
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/accountactionchecks/AccountsActionChecksAPITestsBase.java:144]: Use another constructor

File: src/test/java/org/folio/rest/impl/accountactionchecks/AccountsActionChecksAPITestsBase.java [hunk 15]
```diff
@@ -157,7 +155,7 @@ protected void successfulActionCheckHandlesLongDecimalsCorrectly(ResourceClient
   }
 
   protected void failedActionCheckReturnsInitialRequestedAmount(ResourceClient client) {
-    firstAccount.setRemaining(0.99);
+    firstAccount.setRemaining(new MonetaryValue(new BigDecimal("0.99")));
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/accountactionchecks/AccountsActionChecksAPITestsBase.java:158]: Same here

File: src/test/java/org/folio/test/support/matcher/LogEventMatcher.java [hunk 16]
```diff
@@ -20,44 +20,46 @@
...
 import static org.folio.rest.domain.logs.LogEventPayloadField.SOURCE;
 import static org.folio.rest.domain.logs.LogEventPayloadField.USER_ID;
-import static org.hamcrest.CoreMatchers.allOf;
-import static org.hamcrest.CoreMatchers.anyOf;
-import static org.hamcrest.CoreMatchers.containsString;
-import static org.hamcrest.CoreMatchers.is;
-import static org.hamcrest.CoreMatchers.not;
-import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.CoreMatchers.*;
+import static org.hamcrest.number.IsCloseTo.*;
```
Comment (alexanderkurash) [inline src/test/java/org/folio/test/support/matcher/LogEventMatcher.java:24]: Remove wildcard imports

**Review by OleksandrVidinieiev — APPROVED** [review]

**Pull request merged**